app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/WriteFileAbstraction.java

```diff
@@ -0,0 +1,128 @@
+package com.amaze.filemanager.asynchronous.asynctasks;
+
+import android.content.ContentResolver;
+import android.content.Context;
+import android.os.AsyncTask;
+
+import com.amaze.filemanager.activities.superclasses.ThemedActivity;
+import com.amaze.filemanager.exceptions.ShellNotRunningException;
+import com.amaze.filemanager.exceptions.StreamNotFoundException;
+import com.amaze.filemanager.filesystem.EditableFileAbstraction;
+import com.amaze.filemanager.filesystem.FileUtil;
+import com.amaze.filemanager.filesystem.HybridFileParcelable;
+import com.amaze.filemanager.utils.OnAsyncTaskFinished;
+import com.amaze.filemanager.utils.RootUtils;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.ref.WeakReference;
+
+/**
+ * @author Emmanuel Messulam <emmanuelbendavid@gmail.com>
+ *         on 16/1/2018, at 18:36.
+ */
+
+public class WriteFileAbstraction extends AsyncTask<Void, String, Integer> {
+
+    public static final int NORMAL = 0;
+    public static final int EXCEPTION_STREAM_NOT_FOUND = -1;
+    public static final int EXCEPTION_IO = -2;
+    public static final int EXCEPTION_SHELL_NOT_RUNNING = -3;
+
+    private WeakReference<Context> context;
+    private ContentResolver contentResolver;
+    private EditableFileAbstraction fileAbstraction;
+    private File cachedFile;
+    private OnAsyncTaskFinished<Integer> onAsyncTaskFinished;
+
+    private String dataToSave;
+
+    public WriteFileAbstraction(Context context, ContentResolver contentResolver,
+                                EditableFileAbstraction file, String dataToSave, File cachedFile,
+                                OnAsyncTaskFinished<Integer> onAsyncTaskFinished) {
+        this.context = new WeakReference<>(context);
+        this.contentResolver = contentResolver;
+        this.fileAbstraction = file;
+        this.cachedFile = cachedFile;
+        this.dataToSave = dataToSave;
+        this.onAsyncTaskFinished = onAsyncTaskFinished;
+    }
+
+    @Override
+    protected Integer doInBackground(Void... voids) {
+        try {
+            OutputStream outputStream;
+
+            switch (fileAbstraction.scheme) {
+                case EditableFileAbstraction.SCHEME_CONTENT:
+                    if(fileAbstraction.uri == null) throw new NullPointerException("Something went really wrong!");
+
+                    outputStream = contentResolver.openOutputStream(fileAbstraction.uri);
```

Comment:
You should always catch RuntimeException when you interact with ContentProviders.

ContentProviders are essentially untrusted third-party apps, written by often clueless people. There is little isolation to protect your code from bugs in them.

ContentResolver offers some degree of isolation to you, but it will throw SecurityException if remote ContentProvider denies you access. It will also rethrow most RuntimeExceptions, thrown by remote content provider. E.g. if remote provider accidentally throws NullPointerException, and you don't handle it, *your* app will crash.

Reply:
Okay, will do, thanks.